File: backend-go/internal/application/rbac/menu_service.go
package rbac

import (
	"context"
	"database/sql"
	"strings"
	"time"

	domainrbac "go-backend/internal/domain/rbac"
)

// MenuService 提供菜单管理的用例编排（sys_menu）。
type MenuService struct {
	repo   domainrbac.MenuAdminRepository
	nextID func() int64
	now    func() time.Time
}

func NewMenuService(repo domainrbac.MenuAdminRepository, nextID func() int64) *MenuService {
	return &MenuService{
		repo:   repo,
		nextID: nextID,
		now:    time.Now,
	}
}

func (s *MenuService) ListAll(ctx context.Context) ([]domainrbac.MenuDetail, *Error) {
	list, err := s.repo.ListAll(ctx)
	if err != nil {
		return nil, &Error{Code: "500", Msg: "查询菜单失败"}
	}
	return list, nil
}

func (s *MenuService) Get(ctx context.Context, id int64) (*domainrbac.MenuDetail, *Error) {
	if id <= 0 {
		return nil, &Error{Code: "400", Msg: "ID 参数不正确"}
	}
	item, err := s.repo.Get(ctx, id)
	if err != nil {
		if err == sql.ErrNoRows {
			return nil, &Error{Code: "404", Msg: "菜单不存在"}
		}
		return nil, &Error{Code: "500", Msg: "查询菜单失败"}
	}
	return item, nil
}

type MenuSave struct {
	Type       int16
	Icon       string
	Title      string
	Sort       int32
	Permission string
	Path       string
	Name       string
	Component  string
	Redirect   string
	IsExternal *bool
	IsCache    *bool
	IsHidden   *bool
	ParentID   int64
	Status     int16
}

func (s *MenuService) Create(ctx context.Context, userID int64, req MenuSave) (int64, *Error) {
	m, derr := s.buildMenuForSave(userID, 0, req)
	if derr != nil {
		return 0, derr
	}
	m.ID = s.nextID()
	if m.ID <= 0 {
		return 0, &Error{Code: "500", Msg: "新增菜单失败"}
	}
	now := s.now()
	m.CreateUser = &userID
	m.CreateTime = now
	if err := s.repo.Create(ctx, m); err != nil {
		return 0, &Error{Code: "500", Msg: "新增菜单失败"}
	}
	return m.ID, nil
}

func (s *MenuService) Update(ctx context.Context, userID int64, id int64, req MenuSave) *Error {
	m, derr := s.buildMenuForSave(userID, id, req)
	if derr != nil {
		return derr
	}
	now := s.now()
	m.UpdateUser = &userID
	m.UpdateTime = &now
	if err := s.repo.Update(ctx, m); err != nil {
		return &Error{Code: "500", Msg: "修改菜单失败"}
	}
	return nil
}

func (s *MenuService) Delete(ctx context.Context, ids []int64) *Error {
	if len(ids) == 0 {
		return &Error{Code: "400", Msg: "ID 列表不能为空"}
	}
	if err := s.repo.DeleteCascade(ctx, ids); err != nil {
		return &Error{Code: "500", Msg: "删除菜单失败"}
	}
	return nil
}

func (s *MenuService) buildMenuForSave(userID int64, id int64, req MenuSave) (*domainrbac.Menu, *Error) {
	_ = userID

	if req.Type == 0 {
		req.Type = 1
	}
	title := strings.TrimSpace(req.Title)
	if title == "" {
		return nil, &Error{Code: "400", Msg: "菜单标题不能为空"}
	}

	isExternal := false
	if req.IsExternal != nil {
		isExternal = *req.IsExternal
	}
	isCache := false
	if req.IsCache != nil {
		isCache = *req.IsCache
	}
	isHidden := false
	if req.IsHidden != nil {
		isHidden = *req.IsHidden
	}

	path := strings.TrimSpace(req.Path)
	name := strings.TrimSpace(req.Name)
	component := strings.TrimSpace(req.Component)
	if isExternal {
		if !(strings.HasPrefix(path, "http://") || strings.HasPrefix(path, "https://")) {
			return nil, &Error{Code: "400", Msg: "路由地址格式不正确，请以 http:// 或 https:// 开头"}
		}
	} else {
		if strings.HasPrefix(path, "http://") || strings.HasPrefix(path, "https://") {
			return nil, &Error{Code: "400", Msg: "路由地址格式不正确"}
		}
		if path != "" && !strings.HasPrefix(path, "/") {
			path = "/" + path
		}
		name = strings.TrimPrefix(name, "/")
		component = strings.TrimPrefix(component, "/")
	}

	sort := req.Sort
	if sort <= 0 {
		sort = 999
	}
	status := req.Status
	if status == 0 {
		status = 1
	}

	if id < 0 {
		return nil, &Error{Code: "400", Msg: "ID 参数不正确"}
	}

	return &domainrbac.Menu{
		ID:         id,
		ParentID:   req.ParentID,
		Title:      title,
		Type:       domainrbac.MenuType(req.Type),
		Path:       path,
		Name:       name,
		Component:  component,
		Redirect:   strings.TrimSpace(req.Redirect),
		Icon:       strings.TrimSpace(req.Icon),
		IsExternal: isExternal,
		IsCache:    isCache,
		IsHidden:   isHidden,
		Permission: strings.TrimSpace(req.Permission),
		Sort:       sort,
		Status:     status,
	}, nil
}

File: backend-python/app/services/menu_service.py
"""菜单用例编排（menu 写接口）。"""

from __future__ import annotations

from datetime import datetime
from typing import Any, Optional

from sqlalchemy import delete, select, update
from sqlalchemy.orm import Session

from app.core.id import next_id
from app.db.models.sys_menu import SysMenu
from app.db.models.sys_role_menu import SysRoleMenu
from app.http.response import APIResponse, fail, ok


def _parse_int(v: Any) -> Optional[int]:
    try:
        return int(v)
    except Exception:
        return None


def _build_menu_payload(
    *, menu_id: int, body: dict[str, Any]
) -> tuple[Optional[dict[str, Any]], Optional[APIResponse]]:
    typ = _parse_int(body.get("type") or 0)
    if typ is None:
        return None, fail("400", "请求参数不正确")
    if typ == 0:
        typ = 1

    title = str(body.get("title") or "").strip()
    if title == "":
        return None, fail("400", "菜单标题不能为空")

    is_external = bool(body.get("isExternal")) if body.get("isExternal") is not None else False
    is_cache = bool(body.get("isCache")) if body.get("isCache") is not None else False
    is_hidden = bool(body.get("isHidden")) if body.get("isHidden") is not None else False

    path = str(body.get("path") or "").strip()
    name = str(body.get("name") or "").strip()
    component = str(body.get("component") or "").strip()

    if is_external:
        if path and not (path.startswith("http://") or path.startswith("https://")):
            return None, fail("400", "路由地址格式不正确，请以 http:// 或 https:// 开头")
    else:
        if path.startswith("http://") or path.startswith("https://"):
            return None, fail("400", "路由地址格式不正确")
        if path != "" and not path.startswith("/"):
            path = "/" + path
        name = name.lstrip("/")
        component = component.lstrip("/")

    parent_id = _parse_int(body.get("parentId") or 0)
    if parent_id is None:
        return None, fail("400", "请求参数不正确")

    sort_val = _parse_int(body.get("sort") or 0)
    if sort_val is None:
        return None, fail("400", "请求参数不正确")
    if sort_val <= 0:
        sort_val = 999

    status = _parse_int(body.get("status") or 0)
    if status is None:
        return None, fail("400", "请求参数不正确")
    if status == 0:
        status = 1

    if menu_id < 0:
        return None, fail("400", "ID 参数不正确")

    return (
        {
            "id": menu_id,
            "parent_id": int(parent_id),
            "type": int(typ),
            "title": title,
            "path": path or None,
            "name": name or None,
            "component": component or None,
            "redirect": str(body.get("redirect") or "").strip() or None,
            "icon": str(body.get("icon") or "").strip() or None,
            "is_external": is_external,
            "is_cache": is_cache,
            "is_hidden": is_hidden,
            "permission": str(body.get("permission") or "").strip() or None,
            "sort": int(sort_val),
            "status": int(status),
        },
        None,
    )


def create_menu(*, db: Session, user_id: int, body: dict[str, Any], frontend: Optional[str]) -> APIResponse:
    payload, err = _build_menu_payload(menu_id=0, body=body)
    if err is not None:
        return err

    mid = next_id()
    if mid <= 0:
        return fail("500", "新增菜单失败")

    now = datetime.now()
    try:
        menu_kwargs: dict[str, Any] = {
            "id": mid,
            "title": payload["title"],
            "parent_id": payload["parent_id"],
            "type": payload["type"],
            "path": payload["path"],
            "name": payload["name"],
            "component": payload["component"],
            "redirect": payload["redirect"],
            "icon": payload["icon"],
            "is_external": payload["is_external"],
            "is_cache": payload["is_cache"],
            "is_hidden": payload["is_hidden"],
            "permission": payload["permission"],
            "sort": payload["sort"],
            "status": payload["status"],
            "create_user": int(user_id),
            "create_time": now,
        }
        if frontend is not None:
            menu_kwargs["frontend"] = frontend
        db.add(SysMenu(**menu_kwargs))
        db.commit()
    except Exception:
        db.rollback()
        return fail("500", "新增菜单失败")

    if frontend == "react":
        return ok({"id": str(mid)})
    return ok({"id": mid})


def update_menu(
    *,
    db: Session,
    user_id: int,
    menu_id: int,
    body: dict[str, Any],
    frontend: Optional[str],
) -> APIResponse:
    if menu_id <= 0:
        return fail("400", "ID 参数不正确")

    payload, err = _build_menu_payload(menu_id=menu_id, body=body)
    if err is not None:
        return err

    now = datetime.now()
    try:
        stmt = update(SysMenu).where(SysMenu.id == int(menu_id))
        if frontend is not None:
            stmt = stmt.where(SysMenu.frontend == frontend)
        res = db.execute(
            stmt.values(
                parent_id=payload["parent_id"],
                type=payload["type"],
                title=payload["title"],
                path=payload["path"],
                name=payload["name"],
                component=payload["component"],
                redirect=payload["redirect"],
                icon=payload["icon"],
                is_external=payload["is_external"],
                is_cache=payload["is_cache"],
                is_hidden=payload["is_hidden"],
                permission=payload["permission"],
                sort=payload["sort"],
                status=payload["status"],
                update_user=int(user_id),
                update_time=now,
            )
        )
        if int(getattr(res, "rowcount", 0) or 0) <= 0:
            db.rollback()
            return fail("404", "菜单不存在")
        db.commit()
    except Exception:
        db.rollback()
        return fail("500", "修改菜单失败")

    return ok(True)


def delete_menu_tree(*, db: Session, ids: list[int], frontend: Optional[str]) -> APIResponse:
    if not ids:
        return ok(True)

    stmt = select(SysMenu.id, SysMenu.parent_id)
    if frontend is not None:
        stmt = stmt.where(SysMenu.frontend == frontend)
    rows = db.execute(stmt).all()

    children_of: dict[int, list[int]] = {}
    allowed_ids: set[int] = set()
    for r in rows:
        pid = int(r.parent_id or 0)
        mid = int(r.id)
        allowed_ids.add(mid)
        children_of.setdefault(pid, []).append(mid)

    seed_ids = [mid for mid in ids if mid in allowed_ids]
    if not seed_ids:
        return ok(True)

    seen: set[int] = set()

    def collect(mid: int) -> None:
        if mid in seen:
            return
        seen.add(mid)
        for ch in children_of.get(mid, []):
            collect(ch)

    for mid in seed_ids:
        collect(mid)

    all_ids = sorted(seen)
    if not all_ids:
        return ok(True)

    try:
        db.execute(delete(SysRoleMenu).where(SysRoleMenu.menu_id.in_(all_ids)))
        stmt = delete(SysMenu).where(SysMenu.id.in_(all_ids))
        if frontend is not None:
            stmt = stmt.where(SysMenu.frontend == frontend)
        db.execute(stmt)
        db.commit()
    except Exception:
        db.rollback()
        return fail("500", "删除菜单失败")

    return ok(True)
